src/example.js

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env node
+
+import fs from 'mz/fs';
+import crypto from 'crypto';
+import path from 'path';
+require('source-map-support/register');
+
+import Client from './client';
+import {Controller} from './controller';
+
+/**
+ * This is an example implementation of this library.  It is designed to
+ * demonstrate how this library might be implemented on a client and server.
+ *
+ * NOTE: There is no state maintained about which upload is currently
+ * happening, so the server always assumes only one upload is ever happening
+ * concurrently.  Maintaining this state is a concern of the implementor of
+ * this library.  The UploadID is what should be used here.  This is done
+ * so that this example serves as the minimal example needed to exercise this
+ * library
+ */
+
+
+// This class represents roughly what an api client would look like to the
+// consumer.  Think of the functions other than the constructor as being
+// remotely called methods.  In a CI implementation, this would be the API
+// Client.
+class ServerAPI {
+  constructor(opts) {
+    this.controller = new Controller(/*{region: 'eu-central-1'}*/);
+    this.key = undefined;
+    this.bucket = undefined;
+    this.uploadId = undefined;
+    this.parts = undefined;
+    // NOTE: It's important to remember that while the tags are set for single
+    // part uploads during the creation, the tags sent with the createArtifact
+    // method need to be saved in state for use when running the
+    // completeArtifact method, since multipart uploads set their tags *after*
+    // they're finished
+    // TODO: Verify the above....
+    this.tags = undefined;
+  }
+
+  async createArtifact(opts) {
+    this.bucket = process.env.S3_BUCKET || 'example-remote-s3';
+    this.key = path.basename(opts.filename);
+    this.tags = opts.tags || {};
+    this.parts = opts.parts;
+    this.uploadId = undefined;
+
+    if (opts.parts) {
+      this.uploadId = await this.controller.initiateMultipartUpload({
+        bucket: this.bucket,
+        key: this.key,
+        permissions: {
+          acl: 'public-read',
+        },
+        sha256: opts.sha256,
+        size: opts.size,
+      });
+    }
+
+    let requests;
+    if (opts.parts) {
```

Comment:
having parts for a single part upload means we're duplicating 100% of the information.  There's also a difference between a single part upload and a multipart upload with a single part.